web/src/stream_popover.ts

```diff
@@ -585,6 +586,87 @@ export async function build_move_topic_to_stream_popover(
         $("#move_topic_form .move_messages_edit_topic").trigger("focus");
     }
 
+    // The following logic is correct only when
+    // both message_lists.current.data.fetch_status.has_found_newest
+    // and message_lists.current.data.fetch_status.has_found_oldest are true;
+    // otherwise, we cannot be certain of the correct count.
+    function get_count_of_messages_to_be_moved(
+        selected_option: string,
+        message?: {id: number},
+    ): number {
+        if (selected_option === "change_one") {
+            return 1;
+        }
+        if (selected_option === "change_later") {
+            if (message === undefined) {
+                return 0;
+            }
+            return message_util.get_count_of_messages_in_topic_sent_after_current_message(
+                current_stream_id,
+                topic_name,
+                message.id,
+            );
+        }
+        return message_util.get_messages_in_topic(current_stream_id, topic_name).length;
+    }
+
+    function update_move_messages_count_text(
+        selected_option: string,
+        message?: {id: number},
+    ): void {
+        const message_move_count = get_count_of_messages_to_be_moved(selected_option, message);
+        const has_found_newest = message_lists.current?.data.fetch_status.has_found_newest();
+        const has_found_oldest = message_lists.current?.data.fetch_status.has_found_oldest();
+        const narrowed_stream_id = narrow_state.stream_id();
+        const narrowed_topic_name = narrow_state.topic();
+        const is_topic_narrowed = narrow_state.narrowed_by_topic_reply();
+        const is_stream_narrowed = narrow_state.narrowed_by_stream_reply();
+        const is_same_stream = narrowed_stream_id === current_stream_id;
+        const is_same_topic = narrowed_topic_name === topic_name;
+
+        const can_have_exact_count_in_narrow =
+            (is_stream_narrowed && is_same_stream) ||
+            (is_topic_narrowed && is_same_stream && is_same_topic);
+        let exact_message_count = false;
+        if (selected_option === "change_one") {
+            exact_message_count = true;
+        } else if (
+            selected_option === "change_all" &&
+            can_have_exact_count_in_narrow &&
+            has_found_newest &&
+            has_found_oldest
+        ) {
+            exact_message_count = true;
+        } else if (
+            selected_option === "change_later" &&
+            can_have_exact_count_in_narrow &&
+            has_found_newest
+        ) {
+            exact_message_count = true;
+        }
```

Comment:
This is still somewhat hard to read. Maybe writing it like this would help -
```js
if (selected_option === "change_one") {
    exact_message_count = true;
} else if (can_have_exact_count_in_narrow) {
    const has_found_newest = message_lists.current?.data.fetch_status.has_found_newest();
    const has_found_oldest = message_lists.current?.data.fetch_status.has_found_oldest();

    if (selected_option === "change_later" && has_found_newest) {
        exact_message_count = true;
    }

    if (selected_option === "change_all" && has_found_newest && has_found_oldest) {
        exact_message_count = true;
    }
}
```